File: valuebets.py
"""
Compares model probabilities against bookmaker odds to find value bets.
A value bet exists when your model probability > bookmaker's implied probability.
"""

from config import MIN_EDGE_PCT


def find_value_bets(probs: dict, odds: dict, home: str, away: str) -> list[dict]:
    """
    probs: output of PoissonModel.predict()
    odds:  output of odds.best_odds()
    Returns list of value bet dicts.
    """
    candidates = [
        ("home_win",  "Home Win",  odds.get("home_odds"), probs["home_win"]),
        ("draw",      "Draw",      odds.get("draw_odds"), probs["draw"]),
        ("away_win",  "Away Win",  odds.get("away_odds"), probs["away_win"]),
    ]

    results = []
    for key, label, decimal_odds, model_prob in candidates:
        if not decimal_odds or decimal_odds <= 1.0:
            continue

        implied_prob = 1.0 / decimal_odds
        edge_pct = (model_prob - implied_prob) * 100

        # Kelly criterion: fraction of bankroll to stake
        kelly = (model_prob * decimal_odds - 1.0) / (decimal_odds - 1.0)
        kelly = max(0.0, kelly)

        value = edge_pct >= MIN_EDGE_PCT

        results.append({
            "market": label,
            "model_prob": round(model_prob * 100, 1),
            "implied_prob": round(implied_prob * 100, 1),
            "best_odds": decimal_odds,
            "fair_odds": round(1.0 / model_prob, 2) if model_prob > 0 else None,
            "edge_pct": round(edge_pct, 1),
            "kelly_pct": round(kelly * 100, 1),
            "value": value,
        })

    return results


def print_match_analysis(home: str, away: str, probs: dict,
                          value_bets: list[dict], has_live_odds: bool):
    print(f"\n{'=' * 60}")
    print(f"  {home}  vs  {away}")
    print(f"  xG: {home} {probs['exp_home_goals']} vs {probs['exp_away_goals']} {away}")
    print(f"  Model:  Home {probs['home_win']*100:.1f}%  |  "
          f"Draw {probs['draw']*100:.1f}%  |  "
          f"Away {probs['away_win']*100:.1f}%")

    if not has_live_odds:
        print("  (No bookmaker odds - add THE_ODDS_API_KEY in config.py)")
        return

    print(f"\n  {'Market':<12} {'Model%':>7} {'Impl%':>7} {'B.Odds':>8} "
          f"{'FairOdds':>9} {'Edge%':>7} {'Kelly%':>8}  Verdict")
    print(f"  {'-'*80}")
    for b in value_bets:
        verdict = "[VALUE]" if b["value"] else "-"
        print(f"  {b['market']:<12} {b['model_prob']:>6.1f}% {b['implied_prob']:>6.1f}% "
              f"{b['best_odds']:>8.2f} {b['fair_odds']:>9.2f} "
              f"{b['edge_pct']:>+6.1f}% {b['kelly_pct']:>7.1f}%  {verdict}")

File: lib/valuebets.ts
/**
 * Compares model probabilities against bookmaker odds to find value bets.
 * Port of valuebets.py. A value bet exists when model prob > implied prob.
 */
import { MIN_EDGE_PCT } from "@/lib/config";
import { computeMarketProbs, parseExactScoreOutcome } from "@/lib/markets";
import { round } from "@/lib/util";
import type { Probs } from "@/lib/model";

export interface SimpleOdds {
  home_odds?: number | null;
  draw_odds?: number | null;
  away_odds?: number | null;
}

export interface ValueBet {
  market: string;
  market_group?: string;
  outcome?: string;
  model_prob: number;
  implied_prob: number;
  best_odds: number;
  fair_odds: number | null;
  edge_pct: number;
  kelly_pct: number;
  value: boolean;
}

export function findValueBets(
  probs: Probs,
  odds: SimpleOdds,
  _home: string,
  _away: string
): ValueBet[] {
  const candidates: [string, number | null | undefined, number][] = [
    ["Home Win", odds.home_odds, probs.home_win],
    ["Draw", odds.draw_odds, probs.draw],
    ["Away Win", odds.away_odds, probs.away_win],
  ];

  const results: ValueBet[] = [];
  for (const [label, decimalOdds, modelProb] of candidates) {
    if (!decimalOdds || decimalOdds <= 1.0) continue;

    const impliedProb = 1.0 / decimalOdds;
    const edgePct = (modelProb - impliedProb) * 100;
    const kelly = Math.max(
      0.0,
      (modelProb * decimalOdds - 1.0) / (decimalOdds - 1.0)
    );

    results.push({
      market: label,
      model_prob: round(modelProb * 100, 1),
      implied_prob: round(impliedProb * 100, 1),
      best_odds: decimalOdds,
      fair_odds: modelProb > 0 ? round(1.0 / modelProb, 2) : null,
      edge_pct: round(edgePct, 1),
      kelly_pct: round(kelly * 100, 1),
      value: edgePct >= MIN_EDGE_PCT,
    });
  }
  return results;
}

function makeVbet(
  marketName: string,
  outcomeName: string,
  decimalOdds: number,
  modelProb: number,
  groupName = ""
): ValueBet {
  const impliedProb = 1.0 / decimalOdds;
  const edgePct = (modelProb - impliedProb) * 100;
  const kelly = Math.max(
    0.0,
    (modelProb * decimalOdds - 1.0) / (decimalOdds - 1.0)
  );
  return {
    market: marketName,
    market_group: groupName,
    outcome: outcomeName,
    model_prob: round(modelProb * 100, 1),
    implied_prob: round(impliedProb * 100, 1),
    best_odds: decimalOdds,
    fair_odds: modelProb > 0 ? round(1.0 / modelProb, 2) : null,
    edge_pct: round(edgePct, 1),
    kelly_pct: round(kelly * 100, 1),
    value: edgePct >= MIN_EDGE_PCT,
  };
}

/**
 * True if this bet belongs to the corner model — it must be excluded from the
 * goal-model tab now that corners have their own dedicated tab/model.
 */
export function isCornerBet(v: ValueBet): boolean {
  const mg = v.market_group ?? "";
  const mkt = v.market ?? "";
  return (
    mg === "Szögletek" ||
    mkt.toLowerCase().includes("szöglet") ||
    mg.toLowerCase().includes("szöglet")
  );
}

// Groups to skip entirely (player/card markets — not predictable)
const SKIP_GROUPS = new Set(["Játékosok", "Büntetőlapok"]);

// Market name substrings indicating non-predictable player/card/shot markets
const SKIP_MARKET_SUBSTRINGS = [
  "büntetőlap",
  "sárga lap",
  "piros lap",
  "kaput eltaláló",
  "kapura tartó",
  "szabálytalanság",
  "lesz 11-es",
  "lesz kiállítás",
  "lesz öngól",
  "lesz mesterhármas",
  "kezdőként",
];

export interface MarketOutcome {
  name?: string;
  no?: number;
  fixedOdds?: number | null;
  odds?: number | null;
}
export interface Market {
  name?: string;
  outcomes?: MarketOutcome[];
}
export interface MarketGroup {
  name?: string;
  markets?: Market[];
}

const EXACT_SCORE_MARKETS = new Set([
  "Pontos végeredmény",
  "1. félidő - Pontos eredmény",
  "2. félidő - Pontos eredmény",
]);

export function findAllValueBets(
  marketGroups: MarketGroup[],
  probs: Probs,
  homeHu: string,
  awayHu: string
): ValueBet[] {
  const scoreMatrix = probs.score_matrix ?? [];
  const expH = probs.exp_home_goals;
  const expA = probs.exp_away_goals;

  const results: ValueBet[] = [];
  const seenMarkets = new Set<string>();

  for (const group of marketGroups) {
    const groupName = group.name ?? "";
    if (SKIP_GROUPS.has(groupName)) continue;

    for (const market of group.markets ?? []) {
      const marketName = market.name ?? "";
      if (!marketName) continue;

      const ml = marketName.toLowerCase();
      if (SKIP_MARKET_SUBSTRINGS.some((s) => ml.includes(s))) continue;

      if (seenMarkets.has(marketName)) continue;
      seenMarkets.add(marketName);

      const outcomes = market.outcomes ?? [];

      // Special handling: exact score markets
      if (EXACT_SCORE_MARKETS.has(marketName)) {
        for (const outcome of outcomes) {
          const oddsVal = outcome.fixedOdds || outcome.odds;
          if (!oddsVal || oddsVal <= 1.0) continue;
          const score = parseExactScoreOutcome(
            outcome.name ?? "",
            homeHu,
            awayHu
          );
          if (score === null) continue;
          const [hG, aG] = score;
          if (hG < scoreMatrix.length && aG < (scoreMatrix[0]?.length ?? 0)) {
            const modelProb = scoreMatrix[hG][aG];
            if (modelProb > 0)
              results.push(
                makeVbet(
                  marketName,
                  outcome.name ?? "",
                  oddsVal,
                  modelProb,
                  groupName
                )
              );
          }
        }
        continue;
      }

      // All other markets: use computeMarketProbs
      const outcomeProbs = computeMarketProbs(
        marketName,
        scoreMatrix,
        expH,
        expA,
        homeHu,
        awayHu
      );
      if (!outcomeProbs) continue;

      outcomes.forEach((outcome, idx) => {
        if (!(idx in outcomeProbs)) return;
        const oddsVal = outcome.fixedOdds || outcome.odds;
        if (!oddsVal || oddsVal <= 1.0) return;
        const modelProb = outcomeProbs[idx];
        if (modelProb <= 0) return;
        results.push(
          makeVbet(
            marketName,
            outcome.name ?? "",
            oddsVal,
            modelProb,
            groupName
          )
        );
      });
    }
  }

  // Sort: value bets first, then by edge descending
  results.sort(
    (a, b) =>
      Number(b.value) - Number(a.value) || b.edge_pct - a.edge_pct
  );
  return results;
}
